src/main/java/org/springframework/integration/kafka/listener/KafkaNativeOffsetManager.java

```diff
@@ -0,0 +1,235 @@
+		final Long offset = retryTemplate.execute(new RetryCallback<Long, RuntimeException>() {
+
+			@Override
+			public Long doWithRetry(RetryContext context) throws RuntimeException {
+				context.setAttribute(PARTITION_ATTRIBUTE, partition);
+				Result<Long> result = getOffsetManagerConnection(partition).fetchStoredOffsetsForConsumer(
+						getConsumerId(), partition);
+				checkResultForErrors(result, partition);
+				return result.getResult(partition);
+			}
+
+			@Override
+			public String toString() {
+				return String.format("fetchStoredOffsetsForConsumer(%s, %s)", getConsumerId(), partition);
+			}
+		});
+
+		if (offset != null && offset < 0) {
+			return null;
+		}
+
+		return offset;
+	}
+
+	@Override
+	protected void doUpdateOffset(final Partition partition, final long offset) {
+		retryTemplate.execute(new RetryCallback<Void, RuntimeException>() {
+
+			@Override
+			public Void doWithRetry(RetryContext context) throws RuntimeException {
+				context.setAttribute(PARTITION_ATTRIBUTE, partition);
+				Result<Void> result = getOffsetManagerConnection(partition).commitOffsetsForConsumer(
+						getConsumerId(), Maps.immutable.of(partition, offset).castToMap());
+				checkResultForErrors(result, partition);
+				return null;
+			}
+
+			@Override
+			public String toString() {
+				return String.format("commitOffsetsForConsumer(%s, %s, %s)", getConsumerId(), partition, offset);
+			}
+		});
+	}
+
+	@Override
+	protected void doRemoveOffset(Partition partition) {
+		doUpdateOffset(partition, OffsetAndMetadata.InvalidOffset());
+	}
+
+	@Override
+	public void flush() throws IOException {
+		// this function left intentionally blank
+	}
+
+	@Override
+	public void afterPropertiesSet() throws Exception {
+		if (retryTemplate == null) {
+			retryTemplate = new RetryTemplate();
+			retryTemplate.registerListener(new ResetOffsetManagerBrokerAddressRetryListener());
+			final SimpleRetryPolicy retryPolicy = new SimpleRetryPolicy();
+			retryPolicy.setMaxAttempts(5);
+			retryTemplate.setRetryPolicy(retryPolicy);
+			final ExponentialBackOffPolicy backOffPolicy = new ExponentialBackOffPolicy();
+			backOffPolicy.setInitialInterval(125L);
+			backOffPolicy.setMaxInterval(5000L);
+			backOffPolicy.setMultiplier(2);
+			retryTemplate.setBackOffPolicy(backOffPolicy);
+		}
+	}
+
+	@Override
+	public void close() throws IOException {
+		// this function left intentionally blank
+		zkClient.close();
+	}
+
+	private class ResetOffsetManagerBrokerAddressRetryListener extends RetryListenerSupport {
+
+		@Override
+		public <T, E extends Throwable> void onError(RetryContext context, RetryCallback<T, E> callback, Throwable t) {
+			if (LOGGER.isWarnEnabled()) {
+				LOGGER.warn("Retrying kafka operation [" + callback + "] due to [" + t + "]", t);
+			}
+			final Object partition = context.getAttribute(PARTITION_ATTRIBUTE);
+			if (partition != null) {
+				offsetManagerBrokerAddressCache.remove(partition);
```

Comment:
Minor, of course, but it would be better to make our IDEs happy from those yellow lines :smile: 

![suspicious for map remove](https://cloud.githubusercontent.com/assets/1071590/10707253/a8f348d4-79bd-11e5-8d75-a390653c90bf.jpg)